group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/RangeAssignorTest.java

```diff
@@ -0,0 +1,574 @@
+        Map<Uuid, AssignmentTopicMetadata> topics = Collections.singletonMap(topic1Uuid, new AssignmentTopicMetadata(3));
+        Map<String, AssignmentMemberSpec> members = Collections.singletonMap(
+            consumerA,
+            new AssignmentMemberSpec(
+                    Optional.empty(),
+                    Optional.empty(),
+                    Collections.emptyList(),
+                    Collections.emptyMap())
+            );
+
+        AssignmentSpec assignmentSpec = new AssignmentSpec(members, topics);
+        GroupAssignment groupAssignment = assignor.assign(assignmentSpec);
+
+        assertTrue(groupAssignment.members().isEmpty());
+    }
+
+    @Test
+    public void testOneConsumerNonExistentTopic() {
+        Map<Uuid, AssignmentTopicMetadata> topics = Collections.singletonMap(topic1Uuid, new AssignmentTopicMetadata(3));
+        Map<String, AssignmentMemberSpec> members = Collections.singletonMap(
+            consumerA,
+            new AssignmentMemberSpec(
+                Optional.empty(),
+                Optional.empty(),
+                Collections.singletonList(topic2Uuid),
+                Collections.emptyMap())
+            );
+
+        AssignmentSpec assignmentSpec = new AssignmentSpec(members, topics);
+        GroupAssignment groupAssignment = assignor.assign(assignmentSpec);
+
+        assertTrue(groupAssignment.members().isEmpty());
+    }
+
+    @Test
+    public void testFirstAssignmentTwoConsumersTwoTopicsSameSubscriptions() {
+        Map<Uuid, AssignmentTopicMetadata> topics = new HashMap<>();
+        topics.put(topic1Uuid, new AssignmentTopicMetadata(3));
+        topics.put(topic3Uuid, new AssignmentTopicMetadata(2));
+
+        Map<String, AssignmentMemberSpec> members = new HashMap<>();
+        // Initial Subscriptions are: A -> T1, T3 | B -> T1, T3
+
+        members.put(consumerA, new AssignmentMemberSpec(
+            Optional.empty(),
+            Optional.empty(),
+            Arrays.asList(topic1Uuid, topic3Uuid),
+            Collections.emptyMap())
+        );
+
+        members.put(consumerB, new AssignmentMemberSpec(
+            Optional.empty(),
+            Optional.empty(),
+            Arrays.asList(topic1Uuid, topic3Uuid),
+            Collections.emptyMap())
+        );
+
+        AssignmentSpec assignmentSpec = new AssignmentSpec(members, topics);
+        GroupAssignment computedAssignment = assignor.assign(assignmentSpec);
+
+        Map<Uuid, Set<Set<Integer>>> expectedAssignment = new HashMap<>();
+        // Topic 1 Partitions Assignment
+        mkAssignment(expectedAssignment, topic1Uuid, Arrays.asList(0, 1));
+        mkAssignment(expectedAssignment, topic1Uuid, Collections.singleton(2));
+        // Topic 3 Partitions Assignment
+        mkAssignment(expectedAssignment, topic3Uuid, Collections.singleton(0));
+        mkAssignment(expectedAssignment, topic3Uuid, Collections.singleton(1));
+
+        assertAssignment(expectedAssignment, computedAssignment);
```

Comment:
As I told you offline, I am not a fan of this method. The main issue is that it does not really verify the co-partitioning. Moreover, it does not verify the member ids. I am thinking about two alternatives:

Option 1:
```
Set<String> expectedMembers = mkSet(....);
Set<Map<Uuid, Set<Integer>> expectedAssignments = mkSet(
    mkAssignment(
        mkTopicAssignment(topic1Uuid, 0, 1),
        mkTopicAssignment(topic3Uuid, 0)
    ),
    mkAssignment(
        mkTopicAssignment(topic1Uuid, 2),
        mkTopicAssignment(topic3Uuid, 1)
    ),
);

assertMembers(expectedMembers, computedAssignment)
assertAssignment(expectedAssignment, computedAssignment)
```

Option 2:
We could perhaps use a `TreeMap` instead of an `HashMap` for the members that we pass into the `AssignmentSpec`. The `TreeMap` guarantees the order so the algorithm may be deterministic with this. If it is, we could simply compute the expected `GroupAssignment` and use `assertEquals`.

Reply:
I like the treeMap idea, I wish we thought of this sooner :(

Reply:
Yeah, sorry for this. I only thought about it when I raised this comment.